feat: add custom Google Calendar reminders for booked events

ShashwatPS opened the pull request:
## What does this PR do?

- Fixes #16651 
- Fixes CAL-4716
- /claim #16651

#### Video Demo (if applicable):
https://www.loom.com/share/35ad2beac0714a5d97a56901912c190d?sid=c1606560-2276-4513-8094-ddf5a9cfe7f8

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Run the project locally.
- Install the Google Calendar integration.
- Select your custom reminder (10, 30, or 60 minutes).
- Create an event and make a booking for it.
- Check the event in Google Calendar using the same connected account.

## Checklist  
- [x] I have read the contributing guide  
- [x] My code follows the style guidelines of this project  
- [x] I have commented my code, particularly in hard-to-understand areas  
- [x] I have checked that my changes generate no new warnings

ShashwatPS committed:
Endpoint to update reminder duration

ShashwatPS committed:
Updated schema to adopt global reminder schema

ShashwatPS committed:
Added customReminder on event created for google calendar

ShashwatPS committed:
Bug fixes in update reminder mutation

ShashwatPS committed:
Updated ui for reminder dropdown

ShashwatPS committed:
ui for reminder dropdown

ShashwatPS committed:
bug fixes in reminder dropdown

ShashwatPS committed:
Cleanup

ShashwatPS committed:
Merge branch 'calcom:main' into shashwatps_notification

ShashwatPS committed:
Cleanup

ShashwatPS committed:
tests fix for google calendar event insert

ShashwatPS committed:
Fixes for atoms

ShashwatPS committed:
Merge remote-tracking branch 'origin/main' into shashwatps_notification

ShashwatPS committed:
fixes

kart1ka commented:
@ShashwatPS Thanks for the PR. Could you please fix the failing type check and address the coderabbit suggestions as well?  Please also add tests for this feature.

ShashwatPS committed:
bug and type fixes

ShashwatPS commented on packages/platform/atoms/destination-calendar/DestinationReminderSelector.tsx:
```diff
@@ -0,0 +1,74 @@
+export type DestinationReminderSelectorProps = {
+  destinationCalendar: ConnectedDestinationCalendars["destinationCalendar"];
+  onReminderChange: (value: { credentialId: number; integration: string; defaultReminder: number }) => void;
+};
```
Fixed.

ShashwatPS commented on packages/trpc/server/routers/viewer/calendars/setDestinationReminder.handler.ts:
```diff
@@ -0,0 +1,33 @@
+
+export const setDestinationReminderHandler = async ({ ctx, input }: SetDestinationReminderHandlerOptions) => {
+  const { credentialId, integration, defaultReminder } = input;
+
```
Already addressed in the zod type check.

ShashwatPS commented on packages/trpc/server/routers/viewer/calendars/setDestinationReminder.schema.ts:
```diff
@@ -0,0 +1,11 @@
+  credentialId: z.number(),
+  integration: z.string(),
+  defaultReminder: z.number(),
+});
```
Incorporated a less strict version of this.

ShashwatPS committed:
zod more strict

ShashwatPS committed:
added language support for dropdown

ShashwatPS commented on packages/platform/atoms/destination-calendar/DestinationReminderSelector.tsx:
```diff
@@ -0,0 +1,74 @@
+      { label: "Remind 1 hour before", value: 60 },
+    ],
+    []
+  );
```
Incorporated this.

ShashwatPS committed:
fixing type errors

ShashwatPS commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -581,6 +581,7 @@
   "upload_a_logo": "Upload a logo",
   "upload_logo": "Upload logo",
   "remove_logo": "Remove logo",
+  "remind_minutes_before": "{{count}} minutes before",
```
Not needed.

ShashwatPS committed:
Added test for reminder update

ShashwatPS commented:
@kart1ka  Fixed the type checks, went through the CodeRabbit suggestions and addressed them. Also added tests for the custom reminders in Google Calendar. Please have a look whenever you are free and suggest if anything needs to be changed.

kart1ka committed:
Merge branch 'main' into shashwatps_notification

kart1ka reviewed (approved):
Looks Good. Thanks for the PR.

Pranav-0440 commented:
I would like to pick this up since the existing PR is stale. 
I will create a new PR building on those changes.

syedalamshah commented:
Issue CAL-4716 is fully fixed 
/claim CAL-4716 

All issues are resolved feature is added.

CAL-4716: Custom Google Calendar Notifications
What This PR Does

This PR lets users set custom reminder times for events booked through Cal.com and synced to Google Calendar. Instead of relying on Google Calendar’s default notifications, users can now choose 10, 30, or 60 minutes before the event to receive a notification.

Key Changes

Backend:

Added customCalendarReminder to the DestinationCalendar model (default: 10 minutes)

Created a new tRPC endpoint setDestinationReminder to save users’ reminder preferences

Updated Google Calendar integration to apply custom reminders for both creating and updating events

Both popup and email notifications now respect the selected reminder time

Frontend:

Added DestinationReminderSelector component with a dropdown for 10/30/60 minutes

Integrated the selector into the calendar settings UI

Shows a success toast when preferences are saved

Testing:

Unit tests added for the reminder handler

Verified Google Calendar events now include correct custom reminders

Tested error handling for invalid calendars

How It Works

User connects their Google Calendar

Dropdown appears next to each calendar in settings

User selects a preferred reminder time (10, 30, 60 minutes)

Preference is saved in the database for that calendar

All future bookings automatically include this custom reminder

Rescheduled bookings preserve the chosen reminder

devin-ai-integration[bot] committed:
Merge upstream main into shashwatps_notification - resolve conflicts

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
fix: add missing prisma import in CalendarService

Co-Authored-By: unknown <>
